File: src/main/java/net/guizhanss/fastmachines/FastMachines.java
package net.guizhanss.fastmachines;

import java.util.logging.Level;

import org.bukkit.Bukkit;
import org.bukkit.plugin.java.JavaPlugin;

import io.github.thebusybiscuit.slimefun5.api.SlimefunAddon;
import io.github.thebusybiscuit.slimefun5.implementation.Slimefun;
import io.github.thebusybiscuit.slimefun5.libraries.dough.updater.BlobBuildUpdater;

import org.bstats.bukkit.Metrics;

import net.guizhanss.fastmachines.core.services.ConfigService;
import net.guizhanss.fastmachines.core.services.IntegrationService;
import net.guizhanss.fastmachines.core.services.LocalizationService;
import net.guizhanss.fastmachines.implementation.groups.FMItemGroups;
import net.guizhanss.fastmachines.implementation.items.FMItems;
import net.guizhanss.fastmachines.implementation.listeners.CauldronListener;
import net.guizhanss.fastmachines.implementation.listeners.GravityListener;
import net.guizhanss.fastmachines.implementation.listeners.HopperListener;
import net.guizhanss.fastmachines.implementation.listeners.PlayerProfileListener;
import net.guizhanss.fastmachines.implementation.listeners.SlimefunRegistryListener;
import net.guizhanss.fastmachines.implementation.setup.ResearchSetup;
import net.guizhanss.fastmachines.implementation.tasks.FastMachineTickingTask;
import net.guizhanss.fastmachines.libs.guizhanlib.Scheduler;

/**
 * Main plugin class.
 *
 * Unlike the upstream build this does NOT extend GuizhanLib's {@code AbstractAddon}; it implements the
 * fork's own {@link SlimefunAddon} directly (same approach as the other Java addons in this fork). No
 * Kotlin runtime is involved: the whole plugin is plain Java-8 bytecode, so it loads on every server
 * from 1.8.8 to 26.x.
 */
public class FastMachines extends JavaPlugin implements SlimefunAddon {

    private static final String GITHUB_USER = "ybw0014";
    private static final String GITHUB_REPO = "FastMachines";
    public static final String DEFAULT_LANG = "en-US";

    private static FastMachines instance;
    private static ConfigService configService;
    private static LocalizationService localization;
    private static IntegrationService integrationService;
    private static Scheduler scheduler;

    @Override
    public void onEnable() {
        instance = this;
        scheduler = new Scheduler(this);

        // config
        configService = new ConfigService(this);
        debug("Debug mode is enabled.");

        // localization
        log(Level.INFO, "Loading language...");
        String lang = configService.getLang().getValue();
        localization = new LocalizationService(this, getFile());
        localization.setIdPrefix("FM_");
        localization.addLanguage(lang);
        if (!lang.equals(DEFAULT_LANG)) {
            localization.addLanguage(DEFAULT_LANG);
        }
        log(Level.INFO, "Loaded language " + lang + ".");

        // integrations
        integrationService = new IntegrationService(this);

        // item groups setup
        FMItemGroups.setup();

        // item setup
        FMItems.setup();

        // register per-language item translations (languages/<lang>/items.yml)
        Slimefun.getItemTranslationService().registerTranslations(this);

        // researches setup
        if (configService.getEnableResearches().getValue()) {
            ResearchSetup.setup();
        }

        // listeners & tasks
        setupListeners();
        setupTasks();

        // Metrics setup
        setupMetrics();

        // auto-update
        if (configService.getAutoUpdate().getValue()) {
            autoUpdate();
        }
    }

    @Override
    public void onDisable() {
        Bukkit.getScheduler().cancelTasks(this);
    }

    @Override
    public JavaPlugin getJavaPlugin() {
        return this;
    }

    @Override
    public String getBugTrackerURL() {
        return "https://github.com/" + GITHUB_USER + "/" + GITHUB_REPO + "/issues";
    }

    private void autoUpdate() {
        String version = getDescription().getVersion();
        if (version.startsWith("Dev")) {
            new BlobBuildUpdater(this, getFile(), GITHUB_REPO).start();
        }
    }

    private void setupListeners() {
        new CauldronListener(this);
        new GravityListener(this);
        new HopperListener(this);
        new PlayerProfileListener(this);
        new SlimefunRegistryListener(this);
    }

    private void setupTasks() {
        new FastMachineTickingTask();
    }

    private void setupMetrics() {
        // Consolidated metrics: only start our own bStats if the server opted out (metrics.disable-addon-metrics = false).
        if (Slimefun.getCfg().contains("metrics.disable-addon-metrics") && !Slimefun.getCfg().getBoolean("metrics.disable-addon-metrics")) {
            new Metrics(this, 20046);
        }
    }

    public static FastMachines getInstance() {
        return instance;
    }

    public static ConfigService getConfigService() {
        return configService;
    }

    public static LocalizationService getLocalization() {
        return localization;
    }

    public static IntegrationService getIntegrationService() {
        return integrationService;
    }

    public static SlimefunAddon getAddon() {
        return instance;
    }

    public static Scheduler scheduler() {
        return scheduler;
    }

    public static void log(Level level, String message) {
        instance.getLogger().log(level, message);
    }

    public static void log(Level level, Throwable ex, String message) {
        instance.getLogger().log(level, message, ex);
    }

    public static void debug(String message) {
        if (configService == null || !configService.getDebug().getValue()) {
            return;
        }
        log(Level.INFO, "[DEBUG] " + message);
    }
}

File: src/main/kotlin/net/guizhanss/fastmachines/FastMachines.kt
package net.guizhanss.fastmachines

import io.github.thebusybiscuit.slimefun5.api.SlimefunAddon
import io.github.thebusybiscuit.slimefun5.libraries.dough.updater.BlobBuildUpdater
import net.guizhanss.fastmachines.core.FMRegistry
import net.guizhanss.fastmachines.core.services.ConfigService
import net.guizhanss.fastmachines.core.services.IntegrationService
import net.guizhanss.fastmachines.core.services.LocalizationService
import net.guizhanss.fastmachines.implementation.groups.FMItemGroups
import net.guizhanss.fastmachines.implementation.items.FMItems
import net.guizhanss.fastmachines.implementation.listeners.CauldronListener
import net.guizhanss.fastmachines.implementation.listeners.GravityListener
import net.guizhanss.fastmachines.implementation.listeners.HopperListener
import net.guizhanss.fastmachines.implementation.listeners.PlayerProfileListener
import net.guizhanss.fastmachines.implementation.listeners.SlimefunRegistryListener
import net.guizhanss.fastmachines.implementation.setup.ResearchSetup
import net.guizhanss.fastmachines.implementation.tasks.FastMachineTickingTask
import net.guizhanss.fastmachines.libs.guizhanlib.Scheduler
import org.bstats.bukkit.Metrics
import org.bukkit.Bukkit
import org.bukkit.plugin.java.JavaPlugin
import java.util.logging.Level

/**
 * Main plugin class.
 *
 * Unlike the upstream build this does NOT extend GuizhanLib's `AbstractAddon` (which `implements` the
 * pre-fork `io.github.thebusybiscuit.slimefun4.api.SlimefunAddon` and ships as Java-16 bytecode). It
 * implements the fork's own [SlimefunAddon] directly (same approach as SMG/GeneticChickengineering),
 * and the Kotlin stdlib is now shaded into the jar rather than fetched at runtime via Paper's library
 * loader (which is 1.16.5+ and absent on 1.8.8) - so the `load()`/Libby machinery is gone entirely.
 *
 * [SlimefunAddon] is implemented by a separate [addon] object rather than by this class directly:
 * Bukkit's `PluginBase.getName()` is `final`, so a Kotlin class extending [JavaPlugin] cannot also
 * satisfy `SlimefunAddon`'s conflicting `default getName()`/`getLogger()` (Kotlin, unlike Java, forces
 * an explicit override of the diamond, which a final super method forbids).
 */
class FastMachines : JavaPlugin() {

    override fun onEnable() {
        instance = this
        scheduler = Scheduler(this)
        addon = object : SlimefunAddon {
            override fun getJavaPlugin(): JavaPlugin = this@FastMachines
            override fun getBugTrackerURL(): String = "https://github.com/$GITHUB_USER/$GITHUB_REPO/issues"
        }

        FMRegistry

        // config
        configService = ConfigService(this)
        debug("Debug mode is enabled.")

        // localization
        log(Level.INFO, "Loading language...")
        val lang = configService.lang.value
        localization = LocalizationService(this, file)
        localization.idPrefix = "FM_"
        localization.addLanguage(lang)
        if (lang != DEFAULT_LANG) {
            localization.addLanguage(DEFAULT_LANG)
        }
        log(Level.INFO, "Loaded language $lang.")

        // integrations
        integrationService = IntegrationService(this)

        // item groups setup
        FMItemGroups

        // item setup
        FMItems

        // researches setup
        if (configService.enableResearches.value) {
            ResearchSetup
        }

        // listeners & tasks
        setupListeners()
        setupTasks()

        // Metrics setup
        setupMetrics()

        // auto-update
        if (configService.autoUpdate.value) {
            autoUpdate()
        }
    }

    override fun onDisable() {
        Bukkit.getScheduler().cancelTasks(this)
    }

    private fun autoUpdate() {
        val version = description.version
        if (version.startsWith("Dev")) {
            BlobBuildUpdater(this, file, GITHUB_REPO).start()
        }
    }

    private fun setupListeners() {
        CauldronListener(this)
        GravityListener(this)
        HopperListener(this)
        PlayerProfileListener(this)
        SlimefunRegistryListener(this)
    }

    private fun setupTasks() {
        FastMachineTickingTask()
    }

    private fun setupMetrics() {
        Metrics(this, 20046)
    }

    companion object {

        private const val GITHUB_USER = "ybw0014"
        private const val GITHUB_REPO = "FastMachines"
        const val DEFAULT_LANG = "en-US"

        lateinit var instance: FastMachines
            private set
        lateinit var configService: ConfigService
            private set
        lateinit var localization: LocalizationService
            private set
        lateinit var integrationService: IntegrationService
            private set
        lateinit var addon: SlimefunAddon
            private set
        private lateinit var scheduler: Scheduler

        fun scheduler() = scheduler

        fun log(level: Level, message: String) {
            instance.logger.log(level, message)
        }

        fun log(level: Level, ex: Throwable, message: String) {
            instance.logger.log(level, message, ex)
        }

        fun debug(message: String) {
            if (!Companion::configService.isInitialized || !configService.debug.value) return
            log(Level.INFO, "[DEBUG] $message")
        }
    }
}
